### PR TITLE
Add SQL support for `tinyint` , `smallint`, and `unsigned int variants` 

### LINKED ISSUE #3307: add SQL support for tinyint and unsigned versions of all INTs
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

Currently there is no SQL support for TINYINT, or the unsigned versions of TINYINT, SMALLINT, INT, and BIGINT.
```sql
$ echo "1" > tmp.txt
$ ./target/debug/datafusion-cli.exe
DataFusion CLI v11.0.0
create external table t (c1 tinyint) stored as csv location 'tmp.txt';
NotImplemented("The SQL data type TinyInt(None) is not implemented")
select 1::tinyint;
NotImplemented("Unsupported SQL type TinyInt(None)")
-- or unsigned versions of ints
select 1::int unsigned;
NotImplemented("Unsupported SQL type UnsignedInt(None)")
```
**Describe the solution you'd like**

I'd like SQL support for TINYINT at least. Does DataFusion want SQL support for the unsigned versions of the rest of the INTs? Postgres doesn't support unsigned ints, but sqlparser-rs already has syntax support for it and there are matching Arrow types.

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #3325: add SQL support for unsigned integers
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

Related to PR #3309 which wants to add SQL support for tinyint and unsigned ints. The question is whether we want to support SQL access to underlying data types that Arrow supports. 

Here's a table of integer data type support between Spark, Postgres, DataFusion, Arrow, and Parquet:
image

"SQL support" in this case means only 3 things, as far as I can tell:
* support for the data type in the `CREATE EXTERNAL TABLE x (schema..)` syntax
* support for the data type in the SQL `CAST` function 
* no memory table support via `CREATE TABLE x AS SELECT 1::tinyint unsigned` syntax

Otherwise, the unsigned variants are already supported by DataFusion, because Arrow supports them. Unsigned integers can be in custom table providers, or in a parquet file. 

Without SQL support for these data types, we won't be able to cast to unsigned types from SQL, nor load them from CSV/JSON.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by kmitchener**
# Which issue does this PR close?

Closes #3307 and closes #3325.

 # Rationale for this change

# What changes are included in this PR?

Per the issue, this adds the following SQL datatype support:
* TINYINT
* UNSIGNED {TINYINT, SMALLINT, INT, BIGINT}

so now you can use this in `create external table (c1 tinyint, c2 bigint unsigned) ...` etc. Also available in cast() SQL:

# Are there any user-facing changes?

### CODE REVIEW

**Comment by kmitchener** [comment]
I really messed up the previous PR #3309, this replaces that one. Apologies for that.

File: datafusion/core/tests/sql/mod.rs [hunk 1]
```diff
@@ -621,22 +622,20 @@ async fn register_tpch_csv_data(
 async fn register_aggregate_csv_by_sql(ctx: &SessionContext) {
     let testdata = datafusion::test_util::arrow_test_data();
 
-    // TODO: The following c9 should be migrated to UInt32 and c10 should be UInt64 once
```
Comment (alamb) [inline datafusion/core/tests/sql/mod.rs:624]: 🎉

**Review by alamb — APPROVED** [review]
Looks great -- thank you for sticking with this @kmitchener  as we reached consensus.

cc @liukun4515

**Comment by alamb** [comment]
For anyone else following along, additional work is tracked in https://github.com/apache/arrow-datafusion/issues/3376

**Pull request merged**